remill/BC/DeadStoreEliminator.cpp

```diff
@@ -33,153 +34,314 @@
-      dl(dl_) {}
+      state_ptr(nullptr) {}
 
 void ForwardAliasVisitor::AddInstruction(llvm::Instruction *inst) {
-  curr_wl.insert(inst);
+  if (auto store_inst = llvm::dyn_cast<llvm::StoreInst>(inst)) {
+    llvm::AAMDNodes aamd;
+    store_inst->setAAMetadata(aamd);
+
+  } else if (auto load_inst = llvm::dyn_cast<llvm::LoadInst>(inst)) {
+    llvm::AAMDNodes aamd;
+    load_inst->setAAMetadata(aamd);
+  }
+
+  curr_wl.push_back(inst);
 }
 
-// Iterate through the current worklist, updating the state_offset and state_access_offset
-// according to the instructions in the list. Any instruction that is not
-// currently interpretable (some of its pointers are not yet in the state_offset)
-// is withheld to the next analysis round in the next worklist.
-// Analysis repeats until the current worklist is empty.
-bool ForwardAliasVisitor::Analyze(void) {
+// Iterate through the current worklist, updating the `state_offset` and
+// `state_access_offset` according to the instructions in the list. Any
+// instruction that is not currently interpretable (some of its pointers
+// are not yet in `state_offset`) is withheld to the next analysis round
+// in the next worklist. Analysis repeats until the current worklist is
+// empty or until an error condition is hit.
+bool ForwardAliasVisitor::Analyze(llvm::Function *func) {
+  curr_wl.clear();
+  exclude.clear();
+  state_access_offset.clear();
+  state_offset.clear();
+
+  state_ptr = LoadStatePointer(func);
+  if (!state_ptr) {
+    return false;
+  }
+
+  state_offset.emplace(state_ptr, 0);
+
+  for (auto &block : *func) {
+    for (auto &inst : block) {
+      AddInstruction(&inst);
+    }
+  }
+
+  std::vector<llvm::Instruction *> next_wl;
+  std::vector<llvm::Instruction *> pending_wl;
+  next_wl.reserve(curr_wl.size());
+
   bool progress = true;
-  std::unordered_set<llvm::Instruction *> next_wl;
-  // if any visit makes progress, continue the loop
-  while (!curr_wl.empty() && progress) {
+  bool bump = false;
+
+  while (!curr_wl.empty() && (progress || bump)) {
+    curr_wl.insert(curr_wl.end(), pending_wl.begin(), pending_wl.end());
+    pending_wl.clear();
+
     progress = false;
+
     for (auto inst : curr_wl) {
       switch (visit(inst)) {
         case VisitResult::Progress:
           progress = true;
           break;
+        case VisitResult::Incomplete:
+          pending_wl.push_back(inst);
+          break;
         case VisitResult::NoProgress:
-          next_wl.insert(inst);
+          next_wl.push_back(inst);
+          break;
+        case VisitResult::Ignored:
           break;
         case VisitResult::Error:
           return false;
       }
     }
-    // update curr_wl to the next set of insts
+
     curr_wl.swap(next_wl);
-    // reset upcoming insts
     next_wl.clear();
+
+    if (progress || bump) {
+      bump = false;
+    } else if (!pending_wl.empty()) {
+      bump = true;
+    }
+  }
+
+  // TODO(tim): This condition is triggered a lot.
```

Comment:
Interesting. You'll have to run me through this a bit more; I'm not sure I see what kind of code would allow this to occur.